### PR TITLE
IOS: Reload version fix

### PR DESCRIPTION
**Pull request by sepalani**
This fix homebrews not responding after entering ```IOS_ReloadIOS``` function when the IOS version isn't found.

According to libogc & nintendont source code, this function writes 0 where is located the IOS version and then sends ```IOCTL_ES_LAUNCH```. After that, it waits for the IOS version to be changed in memory (forever when using Dolphin).

Ready to be reviewed & merged.

### CODE REVIEW

**Comment by Parlane** [comment]
@leoetlino one for you :)

**Comment by leoetlino** [comment]
Writing anything other than 0 should be enough to fix them, but technically you're not writing the previous version back, but only the IOS major version -- with the minor version being all zeroes.

Also, are we sure IOS does this when it fails to reload? I'm not entirely convinced it does so (even with the libogc check, which may just be a sanity check more than anything).

**Comment by sepalani** [comment]
My assumptions are based on these lines of code:
- https://github.com/devkitPro/libogc/blob/master/libogc/ios.c#L271
- https://github.com/alfishe/nintendon-t/blob/master/resetstub/ios.c#L318

As far as my understanding goes:
1. During IOS reload they perform some of the calls you mentioned above: ```IOCTL_ES_GETVIEWCNT```, ```IOCTL_ES_GETVIEWS```, ```IOCTL_ES_LAUNCH```. 
2. Then, they write zero instead of the IOS version.
3. Wait for the IOS (major) version to be changed.
4. Check to see if it matches the requested one.

@leoetlino 
Is this behavior the same with the very first old IOSes? Any idea how to fix that, knowing that it doesn't happen on a Wii? The error checking part is done at the end by libogc and returns ```IOS_EMISMATCH``` if an error occurred. The only other ways I can see these fail is if an error is thrown by one of these:
- ```ES_GetNumTicketViews```
- ```ES_GetTicketViews```
- ```ES_LaunchTitleBackground```

IIRC, the homebrews I tried namely uLoader and USB Loader GX wanted to load custom IOSes. I'm almost sure I took my Wii NAND when those where installed. Could it be Dolphin that detected these but replied there were not found during ```IOCTL_ES_LAUNCH```? I'll be happy to close that PR if that's not the proper way to fix it, BTW.

**Comment by leoetlino** [comment]
The check appears to be useless, tbh. It's not as if IOS was going to load another version than the one which is requested, so it looks like it's just a sanity check.

I haven't checked with other IOSes, because it's very likely to be the same, but if you want, just check out ES_Launch and syscall 0x42 & 0x43 (if you're interested: for the former, search for /dev/es, find the resource manager message loop which dispatches IPC requests; for the latter, there's a syscall table at the end of the kernel ELF)

What I think is happening here is that we return valid ticket views (since they do exist in the NAND), then ES_LaunchTitle is called -- and the homebrew assume the launch will always succeed. Are you sure that the homebrew you tested were compiled with the latest version of libogc, one which aborts the launch if libogc's ES_LaunchTitleBackground fails? AFAIK the check was not always present.

**Comment by sepalani** [comment]
It was with a fairly recent version (less than a year) of libogc, at least for [uLoader](https://github.com/sepalani/uloader-sp).

**Comment by sepalani** [comment]
@leoetlino I haven't used my ARM disassembler yet but I found the following strings in the IOS 222:
> Set OS version to %u
> Returning old OS version %u

It seems that other IOSes also have this string:
```sh
$ fgrep -lR "Returning old OS version" *
00000009/content/00000003.app
0000000b/content/00000009.app
0000000c/content/00000001.app
0000000d/content/00000007.app
0000000e/content/0000000b.app
0000000f/content/00000001.app
00000011/content/00000001.app
00000014/content/00000001.app
00000015/content/00000001.app
00000016/content/0000000b.app
000000de/content/00000014.app
000000df/content/0000001e.app
000000e0/content/00000017.app
000000f9/content/0000000e.app
000000fa/content/0000000e.app
```

I still don't know if the libogc does a simple sanity check yet. At least, returning a different version than the one required might be possible, if we assume that this string is indeed used during ```IOS_ReloadIOS```.

Regardless, do you know if we keep track of the current IOS version (I had to use this "hack" to get back the old version)?

**Comment by leoetlino** [comment]
If you look at it with a disassembler, then you won't be making guesses… ;)

It's not possible at all. Those two strings are in a subroutine that's only called by the boot_new_ios_kernel syscall. Here's what it essentially does:

```
u32 set_os_version(const u32 new_version)
{
  const u32 old_version = *((u32*)0x3140);

  if (new_version != 0 && new_version != old_version)
  {
    *((u32*)0x3140) = new_version;
    IOS_FlushDCache(0x3140, 4);
    printf("Set OS version to %u\n", new_version);
  }

  printf("Returning old OS version %u\n", new_version); // sic
  return old_version;
}
```

Moreover, the return value is *not used anywhere*.

**Comment by sepalani** [comment]
@leoetlino Making guesses/assumptions and finding evidences is the key to make reverse engineering easier as long as you don't take anything for granted ;v). I also didn't want to unleash the beast if that's unnecessary. I don't know if you realized what you were saying because that function is a very important evidence. Let me explain.

First, feel free to correct me if I'm wrong since I'm not used to ARM ABI/Assembly and I'm not as familiar with IOSes as you are. Foremost, I [loaded the ELF file](http://wiibrew.org/wiki/ARM_Binaries#ELFLOADER_format) of my cIOS 222 in IDA, ran [this script](http://wiibrew.org/wiki/IOS/Syscall_IDAPython), then I located the ```starlet_syscall_handler``` and rename syscalls according to [Wiibrew](http://wiibrew.org/wiki/IOS/Syscalls).

Rather than focusing on the ```/dev/es``` implication I focused on the function we talked about which is called by ```ios_boot```. I retrieved the string I mentioned earlier that leaded to a subroutine of  ```[[noreturn]] boot_new_ios_kernel(int ios_binary_address, int new_version)``` which shouldn't fail since it's a ```[[noreturn]]``` function that jumps to ```ios_binary_address``` anyway. So, if ```[[noreturn]] IOSError ios_boot(const char* path, u32 flag, u32 version)``` succeeds ```boot_new_ios_kernel``` will be called and can't fail. In other words, ```boot_new_ios_kernel``` will override the IOS version if called, even if you don't have it!

However, ```ios_boot``` can also be called in another function where ```ppc_boot``` can be called alternatively. That function is a subroutine one of the main ones where you can find:```printf("%s\n", "$IOSVersion:  ES: 06/03/09 03:36:55 64M $");```. I also checked ```kernel_set_version``` which overrides the kernel version without further checks if called from ES.

Unless I'm mistaken, we were both wrong. The call to ```boot_new_ios_kernel``` shouldn't fail if you're in the right context:
```C
void syscall_0x43_boot_new_ios_kernel(void *ios_binary_address, unsigned int new_version)
{
  void *var_ios_binary_address;
  int var_new_version;

  var_ios_binary_address = ios_binary_address;
  var_new_version = new_version;
  if ( !sub_FFFF04F4() )  // return dword_FFFFC548[*(_DWORD *)(current_thread_context_addr + 84)];
    boot_new_ios_kernel(var_ios_binary_address, var_new_version);
}

void boot_new_ios_kernel(void *ios_binary_address, int new_version)
{
  int var_new_version;
  void *var_ios_binary_address;

  *((uint32_t*)ios_binary_address + 3) = 0;
  var_new_version = new_version;
  var_ios_binary_address = ios_binary_address;
  sub_FFFF2098();
  *(volatile uint32_t*)0xD80003C = 0;
  sub_FFFF2208();
  sub_FFFF5A44();
  sub_FFFF22D8();
  boot_new_ios_kernel_set_ios_version(var_new_version);
  *(volatile uint32_t*)0x3118 = 0x4000000;
  *(volatile uint32_t*)0x311C = 0x4000000;
  *(volatile uint32_t*)0x3124 = 0x90000800;
  *(volatile uint32_t*)0x3128 = 0x933E0000;
  *(volatile uint32_t*)0x3130 = 0x933E0000;
  *(volatile uint32_t*)0x3120 = 0x93400000;
  *(volatile uint32_t*)0x3134 = 0x93400000;
  *(volatile uint32_t*)0x3148 = 0x93400000;
  *(volatile uint32_t*)0x314C = 0x93420000;
  IOS_FlushDCache(12544, 0x68u);
  kprintf("Updated DDR settings in lomem: 0x%x with 12M settings\n", 12544);
  *(volatile uint32_t*)0xD010018 = 1;
  JUMP(ios_binary_address);
}
```
That is to say, the behaviour of [libogc](https://github.com/devkitPro/libogc/blob/master/libogc/ios.c#L274) makes sense. Especially, if the error value is checked. What I see:
1. [ES_LaunchTitleBackground](https://github.com/devkitPro/libogc/blob/master/libogc/ios.c#L273) -> [IOS_IoctlvRebootBackground](https://github.com/devkitPro/libogc/blob/90b38ee52cb14db11ee961b0e167fda9ed87d090/libogc/es.c#L238) -> [__ipc_syncrequest](https://github.com/devkitPro/libogc/blob/90b38ee52cb14db11ee961b0e167fda9ed87d090/libogc/ipc.c#L1284)
2. Dolphin ```ES::Launch``` returned ```return GetDefaultReply(FS_ENOENT);```.
3. Libogc ret value is 0, not the negative number expected so not an error.

That explains the infinite loop, now I've to find the root cause.

**Comment by leoetlino** [comment]
>So, if [[noreturn]] IOSError ios_boot(const char* path, u32 flag, u32 version) succeeds boot_new_ios_kernel will be called and can't fail. In other words, boot_new_ios_kernel will override the IOS version if called, even if you don't have it!

That's correct. However, boot_new_ios_kernel (which corresponds to syscall 0x43) is never called directly from ES. ES only uses either syscall 0x41 (ppc_boot) or 0x42 (ios_boot). The latter indirectly calls boot_new_ios_kernel, but only after it's made sure the boot cannot fail, as I've said and as you've verified.

>However, ios_boot can also be called in another function where ppc_boot can be called alternatively. That function is a subroutine one of the main ones where you can find:

Not sure I follow. The subroutine which calls either ppc_boot or ios_boot is in ES_Launch, not in es_main (the huge function with the version print).

>I also checked kernel_set_version which overrides the kernel version without further checks if called from ES.

Yes, that's normal. That's how the new kernel knows where to load the modules from (syscall 0x4d in es_main). Unless I missed something, kernel_set_version (syscall 0x4c) is not called directly from ES, or internally from the kernel, though.

> That is to say, the behaviour of libogc makes sense. Especially, if the error value is checked.

Imo, this doesn't justify the check in libogc. I think the issue here is that this is a case which cannot happen on the real hardware: on a Wii, if the an IOS exists (GetTicketView ret = 0), then the launch is pretty much always going to succeed. I guess the ES_Launch failure is throwing it off, especially if it's running the request in the background.

**Comment by sepalani** [comment]
@leoetlino I finally found the reason why the Dolphin [response got erased](https://github.com/devkitPro/libogc/blob/90b38ee52cb14db11ee961b0e167fda9ed87d090/libogc/ipc.c#L421). TBH, I really think that's an issue because that case is handled by the IOSes and return an error code and the fact that we can't rely on it seems strange. Might be Dolphin's fault or developers' fault and by extension libogc's fault. As of right now, I assume that's a timing issue.

NTS: Debugging Dolphin that's great but leads to many issues, memory breakpoints and watches don't work if we're in Dolphin emitted instructions (in VS2015). So, debugging Dolphin with its own debugger is more efficient (as long as you know that the problem is there ><'). Though, dealing with interruptions isn't fun since you're teleported here and there, as well.

In short, when the program [sleeps](https://github.com/devkitPro/libogc/blob/90b38ee52cb14db11ee961b0e167fda9ed87d090/libogc/ipc.c#L711) we get teleported to various places due to interruptions and more specifically the [__ipc_interrupthandler](https://github.com/devkitPro/libogc/blob/90b38ee52cb14db11ee961b0e167fda9ed87d090/libogc/ipc.c#L434) and it seems that some functions got inlined. The [__ipc_ackhandler]() is overriding the IOS request result while we're sleeping. That's why I'm wondering if we're handling properly the ack mechanism and that's libogc's fault or if it's Dolphin.

**Comment by leoetlino** [comment]
Sounds more like libogc's fault, as we seem to handle acknowledgments properly (though the second ack is handled in a strange way...)

IOS acks after receiving the IPC request, and then acks a second time after the new IOS kernel has booted (as part of its IPC initialisation). If the launch fails, IOS will still ack the first time, but then it'll reply to the IPC request through the mailbox. As I've said, the latter typically doesn't happen on the real hardware, so it's entirely possible this is a bug in libogc that no one was able to reproduce and notice because ES_Launch always works on a Wii.

If you have a Wii and some non-critical IOS, you could try this. Delete the kernel for e.g. IOS222 (be careful not to delete a shared content...), then write a homebrew that launches that IOS, and see what happens. (IOS should return -106.)

**Comment by sepalani** [comment]
I've a Wii NAND dump and BootMii installed on boot2 so I should be fine. I should borrow a screen somewhen and meanwhile write a bunch of hardware tests. Do you think we should always allow ES_Launch to succeed and put a PanicAlert to say that it might lead to strange behavior if we don't emulate that specific IOS or stay with that behaviour that isn't handled by libogc (apparently?)?
